### PR TITLE
[Python API] Fix python api for bytecode

### PR DESCRIPTION
**Pull request by maxbartel**
This fixes an issue seen in https://github.com/iree-org/iree/pull/17278. MLIR bytecode can have a zero as the last byte depending on the IR. If this happens by accident parsing will fail. This PR checks if a buffer begins with the _magic number_ from https://mlir.llvm.org/docs/BytecodeFormat/#magic-number and sets zero-termination to false if it does.

However I could not generate MLIR bytecode that ends with a zero other than the one in the PR. Even if I save the IR used in textual format and parse it again, some additional bytes are appended and it is not zero terminated anymore. I think this is because of some behavior in TF.

Still after reading a bit through https://mlir.llvm.org/docs/BytecodeFormat I think the check implemented in this PR is the correct behavior for bytecode.

### CODE REVIEW

File: compiler/bindings/python/test/api/testdata/generate_mlir_bytecode.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    if not iree.compiler.tools.tflite.is_available():
+        return
+    this_dir = os.path.dirname(__file__)
+    path = os.path.join(this_dir, "..", "..", "tools", "testdata", "tflite_sample.fb")
+    bytecode = iree.compiler.tools.tflite.compile_file(path, import_only=True)
+    with open(
+        os.path.join(this_dir, "bytecode_zero_terminated_testfile.bc"), "wb"
+    ) as file:
+        file.write(bytecode)
+
```
Comment (maxbartel) [inline compiler/bindings/python/test/api/testdata/generate_mlir_bytecode.py:43]: If someone has an easier way to generate zero terminated bytecode, I am more than happy to remove this!

**Comment by stellaraccident** [comment]
Ugh, that broken llvm API strikes again. Thanks for the fix. Will review in detail soon, since this has been a trouble area and I need to focus on it

File: compiler/bindings/python/iree/compiler/api/ctypes_dl.py [hunk 2]
```diff
@@ -179,6 +179,16 @@ def _initializeGlobalCL(*cl_args: str):
...
 
 
+def _is_null_terminated(view: memoryview):
+    return view.nbytes > 0 and view[-1] == 0
+
+
+def _is_mlir_bytecode(view: memoryview):
+    """Compares the first 4 bytes of the view against the magic number 4d4cef52.
+    See https://mlir.llvm.org/docs/BytecodeFormat/#magic-number for more info."""
+    return view[:4].hex() == "4d4cef52"
```
Comment (stellaraccident) [inline compiler/bindings/python/iree/compiler/api/ctypes_dl.py]: Need to check the length. `return len(view) >= 4 and ...`

File: compiler/bindings/python/test/api/testdata/generate_mlir_bytecode.py (hunk 1 above)
Comment (stellaraccident) [reply compiler/bindings/python/test/api/testdata/generate_mlir_bytecode.py:43]: I am pretty sure it is not valid and that tf is doing bad string manipulation. We can still patch it, but yeah, probably won't be easy to test

**Review by stellaraccident — APPROVED** [review]
I'm not going to claim I completely understand the nuance here, but let's land the patch. The way this was built on the llvm side was wrong a long time ago, and this badness all stems from that.

**Comment by maxbartel** [comment]
Force pushed because of DCO

**Pull request merged**